semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -446,136 +454,117 @@ let rec m_name a b =
-             (B.IdQualified { nameinfo with name_info = B.empty_id_info () })
-      >||>
-      (* try this time by replacing the qualifier by the resolved one *)
-      let new_qualifier =
-        match List.rev dotted with
-        | [] -> raise Impossible
-        | _x :: xs -> List.rev xs |> Common.map (fun id -> (id, None))
-      in
-      m_name a
-        (B.IdQualified
-           {
-             nameinfo with
-             name_middle = Some (B.QDots new_qualifier);
-             name_info = B.empty_id_info ();
-           })
-  (* semantic! try to handle open in OCaml by querying LSP! The
-   * target code is using an unqualified Id possibly because of some open!
-   *)
-  | G.IdQualified { name_last = ida, None; _ }, B.Id (idb, _infob)
-    when fst ida = fst idb -> (
-      match !Hooks.get_def idb with
-      | None -> fail ()
-      | Some file ->
-          let m = module_name_of_filename file in
-          let t = snd idb in
-          pr2_gen m;
-          let _n = H.name_of_ids [ (m, t); idb ] in
-          (* retry with qualified target *)
-          (* m_name a n *)
-          return ())
-  (* boilerplate *)
-  | ( _,
-      G.IdQualified
-        ({
-           name_info =
-             {
-               B.id_resolved =
-                 {
-                   contents =
-                     Some
-                       ( (( B.ImportedEntity dotted
-                          | B.ImportedModule (B.DottedName dotted)
-                          | B.ResolvedName (dotted, _) ) as resolved),
-                         _sid );
-                 };
-               _;
-             };
-           _;
-         } as b1) ) -> (
-      try_parents dotted
-      >||> try_alternate_names resolved
-      >||>
-      match a with
-      | IdQualified a1 -> m_name_info a1 b1
-      | Id _ -> fail ())
-  | G.IdQualified a1, B.IdQualified b1 -> m_name_info a1 b1
-  | G.Id _, _
-  | G.IdQualified _, _ ->
-      fail ()
+      (match a with
+      | B.Id ((str, tok), _info) when MV.is_metavar_name str ->
+          (* If `is_distributed` is true, then we got the target `IdQualified` from a
+             `ResolvedName`.
+             Such an identifier has a nonsensical range, because it may have been
+             constructed from identifiers from arbitrary areas, such as imports or
+             classes.
+
+             If the resolved name is `A.foo`, in
+             the following code:
+             ```
+             class A:
+               def foo():
+                 pass
+             ```
+             then the pattern metavariable will match the range of the text
+             ```
+             A:
+             def foo
+             ```
+             because it is matching everything between `A` and `foo`.
+
+             So we should not permit this match.
+          *)
+          if is_distributed then fail () else envf (str, tok) (MV.N b)
+      | B.Id _ -> fail ()
+      | B.IdQualified a1 -> m_name_info a1 nameinfo)
```

Comment:
Previously this was combined with an `>||>` and now it's combined with a `>!>`, could add a comment explaining why?